File: MountLicense/src/main/java/org/cubexmc/mountlicense/listener/RegistrationListener.java
package org.cubexmc.mountlicense.listener;

import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerInteractEntityEvent;
import org.bukkit.inventory.EquipmentSlot;
import org.bukkit.inventory.ItemStack;
import org.cubexmc.mountlicense.MountLicensePlugin;
import org.cubexmc.mountlicense.service.PdcKeys;
import org.cubexmc.mountlicense.service.RegistryService;
import org.cubexmc.mountlicense.service.RegistryService.Result;
import org.bukkit.persistence.PersistentDataType;

public class RegistrationListener implements Listener {

    private final MountLicensePlugin plugin;
    private final RegistryService registryService;
    private final PdcKeys keys;

    public RegistrationListener(MountLicensePlugin plugin, RegistryService registryService, PdcKeys keys) {
        this.plugin = plugin;
        this.registryService = registryService;
        this.keys = keys;
    }

    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    public void onInteractEntity(PlayerInteractEntityEvent event) {
        if (event.getHand() != EquipmentSlot.HAND) return;

        Player player = event.getPlayer();
        if (!player.isSneaking()) return;

        ItemStack hand = player.getInventory().getItemInMainHand();
        if (!isLicenseRole(hand)) return;

        Entity target = event.getRightClicked();
        double distance = player.getLocation().distance(target.getLocation());
        if (distance > plugin.configManager().getMaxInteractDistance()) return;

        Result result = registryService.tryRegister(player, target, hand);
        if (result == Result.SUCCESS || result != Result.NO_PROFILE) {
            event.setCancelled(true);
        }
    }

    private boolean isLicenseRole(ItemStack item) {
        if (item == null) return false;
        if (item.getItemMeta() == null) return false;
        String role = item.getItemMeta().getPersistentDataContainer()
                .get(keys.itemRole(), PersistentDataType.STRING);
        return PdcKeys.ITEM_ROLE_LICENSE.equals(role);
    }
}

File: MountLicense/src/main/java/org/cubexmc/mountlicense/listener/RegistrationListener.kt
package org.cubexmc.mountlicense.listener

import org.bukkit.entity.Entity
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerInteractEntityEvent
import org.bukkit.inventory.EquipmentSlot
import org.bukkit.inventory.ItemStack
import org.bukkit.persistence.PersistentDataType
import org.cubexmc.mountlicense.MountLicensePlugin
import org.cubexmc.mountlicense.service.PdcKeys
import org.cubexmc.mountlicense.service.RegistryService

class RegistrationListener(
    private val plugin: MountLicensePlugin,
    private val registryService: RegistryService,
    private val keys: PdcKeys,
) : Listener {
    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    fun onInteractEntity(event: PlayerInteractEntityEvent) {
        if (event.hand != EquipmentSlot.HAND) return

        val player: Player = event.player
        if (!player.isSneaking) return

        val hand: ItemStack = player.inventory.itemInMainHand
        if (!isLicenseRole(hand)) return

        val target: Entity = event.rightClicked
        val distance = player.location.distance(target.location)
        if (distance > plugin.configManager().getMaxInteractDistance()) return

        val result = registryService.tryRegister(player, target, hand)
        if (result == RegistryService.Result.SUCCESS || result != RegistryService.Result.NO_PROFILE) {
            event.isCancelled = true
        }
    }

    private fun isLicenseRole(item: ItemStack?): Boolean {
        if (item == null) return false
        val meta = item.itemMeta ?: return false
        val role = meta.persistentDataContainer.get(keys.itemRole(), PersistentDataType.STRING)
        return PdcKeys.ITEM_ROLE_LICENSE == role
    }
}
